modules/beefy/src/utils.rs

```diff
@@ -0,0 +1,361 @@
+>(
+	validators: V,
+) -> BridgedMmrHash<T, I> {
+	let merkle_leafs = validators
+		.cloned()
+		.map(BridgedBeefyValidatorIdToMerkleLeaf::<T, I>::convert)
+		.collect::<Vec<_>>();
+	merkle_root::<BridgedMmrHashing<T, I>, _>(merkle_leafs)
+}
+
+fn verify_validator_set<T: Config<I>, I: 'static>(
+	authority_set: &BridgedBeefyAuthoritySet<T, I>,
+	validator_set: &BridgedBeefyValidatorSet<T, I>,
+) -> Result<(), Error<T, I>> {
+	ensure!(validator_set.id() == authority_set.id, Error::<T, I>::InvalidValidatorSetId);
+	ensure!(
+		validator_set.len() == authority_set.len as usize,
+		Error::<T, I>::InvalidValidatorSetLen
+	);
+
+	// Ensure that the authority set that signed the commitment is the expected one.
+	let validator_set_root = get_validators_mmr_root::<T, I, _>(validator_set.validators().iter());
+	ensure!(validator_set_root == authority_set.root, Error::<T, I>::InvalidValidatorSetRoot);
+
+	Ok(())
+}
+
+/// Number of correct signatures, required from given validators set to accept signed
+/// commitment.
+///
+/// We're using 'conservative' approach here, where signatures of `2/3+1` validators are
+/// required..
+pub(crate) fn signatures_required<T: Config<I>, I: 'static>(validators_len: usize) -> usize {
+	validators_len - validators_len.saturating_sub(1) / 3
+}
+
+fn verify_signatures<T: Config<I>, I: 'static>(
+	commitment: &BridgedBeefySignedCommitment<T, I>,
+	validator_set: &BridgedBeefyValidatorSet<T, I>,
+) -> Result<(), Error<T, I>> {
+	ensure!(
+		commitment.signatures.len() == validator_set.len(),
+		Error::<T, I>::InvalidCommitmentSignaturesLen
+	);
+
+	// Ensure that the commitment was signed by enough authorities.
+	let msg = commitment.commitment.encode();
+	let mut missing_signatures = signatures_required::<T, I>(validator_set.len());
+	for (idx, (authority, maybe_sig)) in
+		validator_set.validators().iter().zip(commitment.signatures.iter()).enumerate()
+	{
+		if let Some(sig) = maybe_sig {
+			if BeefyVerify::<BridgedBeefyCommitmentHasher<T, I>>::verify(sig, &msg, authority) {
+				missing_signatures = missing_signatures.saturating_sub(1);
+				if missing_signatures == 0 {
+					break
+				}
+			} else {
+				log::debug!(
+					target: LOG_TARGET,
+					"Signed commitment contains incorrect signature of validator {} ({:?}): {:?}",
+					idx,
+					authority,
+					sig,
+				);
+			}
+		}
+	}
+	ensure!(missing_signatures == 0, Error::<T, I>::NotEnoughCorrectSignatures);
+
+	Ok(())
+}
+
+/// Extract MMR root from commitment payload.
+fn extract_mmr_root<T: Config<I>, I: 'static>(
+	commitment: &BridgedBeefySignedCommitment<T, I>,
+) -> Result<BridgedMmrHash<T, I>, Error<T, I>> {
+	commitment
+		.commitment
+		.payload
+		.get_decoded(&bp_beefy::MMR_ROOT_PAYLOAD_ID)
+		.ok_or(Error::MmrRootMissingFromCommitment)
+}
+
+pub(crate) fn verify_commitment<T: Config<I>, I: 'static>(
+	commitment: &BridgedBeefySignedCommitment<T, I>,
+	authority_set: &BridgedBeefyAuthoritySet<T, I>,
```

Comment:
Is there a difference between authorities and validators here? If not, then what `authority_set` means and how it is different from `validator_set`? Let's probably stick to same term everywhere - e.g. use authorities or validators everywhere?

Reply:
Sorry, this names are confusing. Added more context about this in a different comment.